Source/Core/Core/HW/WiimoteEmu/Extension/Drums.cpp

```diff
@@ -77,20 +77,64 @@ void Drums::Update()
     drum_data.sy = static_cast<u8>((stick_state.y * STICK_RADIUS) + STICK_CENTER);
   }
 
-  // TODO: Implement these:
-  drum_data.which = 0x1F;
-  drum_data.none = 1;
-  drum_data.hhp = 1;
-  drum_data.velocity = 0xf;
-  drum_data.softness = 7;
-
   // buttons
   m_buttons->GetState(&drum_data.bt, drum_button_bitmasks.data());
 
   // pads
   m_pads->GetState(&drum_data.bt, drum_pad_bitmasks.data());
 
-  // flip button bits
+  // Currently using a fixed "softness".
+  constexpr u8 VELOCITY_SOFTNESS = 3;
+
+  // Send no velocity data by default.
+  drum_data.velocity_id = u8(VelocityID::None);
+  drum_data.no_velocity_data_0b11 = 0b11;
+  drum_data.always_0b11 = 0b11;
+  drum_data.no_velocity_data = 1;
+  drum_data.softness = 7;
+
+  // Clear velocity-sent state for released pads.
+  m_velocity_data_sent &= drum_data.bt;
+
+  // If a pad is hit send velocity data no more than once.
+  for (auto pad_mask : drum_pad_bitmasks)
+  {
+    if ((drum_data.bt & pad_mask) && !(m_velocity_data_sent & pad_mask))
```

Comment:
I wonder about the naming of "mask" - if this is really a bitmask, can multiple pads be hit at the same time? I imagine this would be a difficult to hit at exactly the same input-frame, but probably causes a missed input somewhere once in a blue moon (in a TAS perhaps? If anyone ever attempted to do one of those...)

Reply:
Each song has maybe a few hundred points where you're supposed to hit two pads at once (at least if you're playing on Expert), so the chance that the two hits end up being close enough to each other at least once is not that small in the long run.

Reply:
`VelocityID` kinda looks like a bitmask as well (although a very strange one), so I wouldn't be surprised if they could just be mashed together into a single report.

I don't suppose we have someone with hardware and some homebrew to try this out?

Reply:
The "pad" (drum pad) bytes are a mask, simultaneous hits are ORed together.
The velocity ID is not.

The way I coded it is if pads are simultaneously hit the velocity for each will be sent one after the other (in separate input frames) so long as the inputs are held long enough (N frames for N simultaneous inputs).

So if 2 pads are simultaneously pressed and held for exactly 1 input frame then velocity data will not be sent for one of the pads.

I do not know how the real hardware functions.

I should mention much of the game functions without sending any velocity data.
There seem to be certain menu areas that strangely require it though.

Reply:
Makes sense (re. "mask" naming).

Any other ideas on simultaneous input? I cannot imagine that real hardware would just drop one of the inputs in that case.

Reply:
It doesn't really "drop the input". It just doesn't send velocity data. (which much of the game seems to be fine with.) But I can make the code send velocity data for every hit no matter what.